tools/setup

```diff
@@ -1,51 +1,98 @@
 #!/bin/sh
 
 ROOT=$(dirname $0)/..
-
 NVM_VERSION="v0.34.0"
-
-BAZEL_DEB="https://github.com/bazelbuild/bazel/releases/download/0.29.0/bazel_0.29.0-linux-x86_64.deb"
-BAZEL_PKG_NAME=$(basename "$BAZEL_DEB")
+BAZEL_VERSION="0.29.0"
 
 YLW="\033[33m"
 RED="\033[91m"
 GRN="\033[92m"
 MAG="\033[95m"
 BLD="\033[1m"
 END="\033[0m"
-
 CMD="$BLD$YLW"
 
+warn() {
+    echo "$RED$BLD$1$END"
+}
 
 fail() {
-    echo "$RED$BLD$1$END"
+    warn $1
     exit 1
 }
 
 status() {
     echo "$MAG$1$END"
 }
 
+# Detect Platform
+case $(uname | tr '[:upper:]' '[:lower:]') in
+  linux*)
+    OS_NAME=linux
+    BAZEL_DEP="https://github.com/bazelbuild/bazel/releases/download/$BAZEL_VERSION/bazel_$BAZEL_VERSION-installer-linux-x86_64.sh"
+    ;;
+  darwin*)
+    OS_NAME=osx
+    BAZEL_DEP="https://github.com/bazelbuild/bazel/releases/download/$BAZEL_VERSION/bazel-$BAZEL_VERSION-installer-darwin-x86_64.sh"
+    ;;
+  msys*)
+    OS_NAME=windows
+    BAZEL_DEP="https://github.com/bazelbuild/bazel/releases/download/$BAZEL_VERSION/bazel-$BAZEL_VERSION-windows-x86_64.exe"
+    # TODO(alxr) investigate windows compatability
+    fail "Windows is currently unsupported. Please follow install instructions in the README."
+    ;;
+  *)
+    OS_NAME=notset
+    warn "Can't detect host OS"
+    ;;
+esac
+
+BAZEL_PKG_NAME=$(basename "$BAZEL_DEP")
+
+status "0. Pre-requisites"
+if [ "$OS_NAME" == "linux" ]; then
+  status "0.1 Installing pre-requisites for Bazel"
+  sudo apt-get install pkg-config zip g++ zlib1g-dev unzip python3
+elif [ "$OS_NAME" == "osx" ]; then
+  status "0.1. Checking for Command Line Tools"
+  xcode-select --version || xcode-select --install || fail "Unable to install Command Line Tools"
+fi
+
 status "1. Install nvm"
-curl -o- "https://raw.githubusercontent.com/creationix/nvm/$NVM_VERSION/install.sh" | sh || \
+curl -L -o- "https://raw.githubusercontent.com/creationix/nvm/$NVM_VERSION/install.sh" | sh || \
     fail "Failed to download nvm version $NVM_VERSION."
 
 status "1.1 ensure nvm is in your current process"
 . ~/.nvm/nvm.sh || source ~/.nvm/nvm.sh
 
 status "2. Install node"
-(cd $ROOT && nvm install && nvm use) || fail 'Failed to install target node version.'
+pushd .
+cd $ROOT && nvm install && nvm use || fail 'Failed to install target node version.'
+popd
```

Comment:
How come you're using pushd/popd here but are using cd in parens `(cd ...)` below?

Reply:
I found this out via trial and error: Wrapping the command in parents won't change the intended node version. `nvm use` will only effect node within the subshell. 

([Reference](https://unix.stackexchange.com/questions/26063/how-are-parentheses-interpreted-at-the-command-line))

Reply:
I use `cd` below since they create side effects in the filesystem, and we don't permanently want to change the directory to root.

Reply:
.. ended up removing popd / pushd to support more posix platforms.